File: framework/src/main/java/com/github/mrmitew/bodylog/framework/common/view/InjectableActivity.java
package com.github.mrmitew.bodylog.framework.common.view;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import com.github.mrmitew.bodylog.framework.AndroidApplication;
import com.github.mrmitew.bodylog.framework.di.activity.HasActivitySubcomponentBuilders;

public abstract class InjectableActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setupActivityComponent();
        super.onCreate(savedInstanceState);
    }

    protected void setupActivityComponent() {
        injectMembers(AndroidApplication.get(this));
    }

    protected abstract void injectMembers(HasActivitySubcomponentBuilders
                                                  hasActivitySubcomponentBuilders);
}

File: framework/src/main/java/com/github/mrmitew/bodylog/framework/common/view/InjectableActivity.kt
package com.github.mrmitew.bodylog.framework.common.view

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

import com.github.mrmitew.bodylog.framework.AndroidApplication
import com.github.mrmitew.bodylog.framework.di.activity.HasActivitySubcomponentBuilders

abstract class InjectableActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        setupActivityComponent()
        super.onCreate(savedInstanceState)
    }

    protected abstract fun injectMembers(hasActivitySubcomponentBuilders: HasActivitySubcomponentBuilders)

    private fun setupActivityComponent() {
        injectMembers(AndroidApplication.getActivitySubcomponentBuilders(application))
    }
}